### PR TITLE
Correctly include port in host header

### PR DESCRIPTION
**Pull request by robalar**
The correct form of the Host header is ``<host>:<port>`.  If no port is included, the default port for the service requested is implied (e.g., 443 for an HTTPS URL, and 80 for an HTTP URL).

reqwest's `host_str()` does not return a port, even if a non-default one is set, there for we need to add it.

### CODE REVIEW

File: src/reqwest_impls.rs [hunk 1]
```diff
@@ -52,7 +58,13 @@ impl RequestLike for reqwest::blocking::Request {
...
 impl ClientRequestLike for reqwest::blocking::Request {
     fn host(&self) -> Option<String> {
-        self.url().host_str().map(Into::into)
+        let host = self.url().host_str();
+        let port = self.url().port();
+
+        host.map(|host| match port {
+            Some(port) => format!("{}:{}", host, port),
+            None => host.into(),
+        })
```
Comment (thomassa) [inline src/reqwest_impls.rs:67]: This is the same for `reqwest::Request` and `reqwest::blocking::Request`.
Maybe deduplicate as
```rust
fn host(&self) -> Option<String> {
    host_and_port(self.url())
}
```
with `fn host_and_port` having body
```
url.host_str().map(|host| match url.port() {
    Some(port) => format!("{}:{}", host, port),
    None => host.into(),
})
```

File: src/reqwest_impls.rs [hunk 2]
```diff
@@ -102,6 +114,29 @@ mod tests {
...
+                    .format("%a, %d %b %Y %T GMT")
+                    .to_string(),
+            )
+            .body(&br#"{ "x": 1, "y": 2}"#[..])
+            .build()
+            .unwrap();
+
+        let with_sig = without_sig.signed(&config).unwrap();
+        assert_eq!(with_sig.headers().get("host").unwrap(), "localhost:8080");
+    }
```
Comment (thomassa) [inline src/reqwest_impls.rs:140]: Is it worth adding a similar `assert_eq` to the existing `it_works` test, to test that the newly-changed `host` functions work aright when no port is specified?

Also... should we test both `reqwest::Request` and `reqwest::blocking::Request`?

Comment (robalar) [reply src/reqwest_impls.rs:140]: yeah i meant to test this over both whoops

**Comment by thomassa** [comment]
I think the third commit ("Remove unused export") should be squashed into "Remove use of deprecated chrono functions".

"Therefore" is just one word.

**Comment by thomassa** [comment]
Oops, closed in error. Sorry!

File: src/reqwest_impls.rs [hunk 3]
```diff
@@ -20,7 +20,13 @@ impl RequestLike for reqwest::Request {
 
 impl ClientRequestLike for reqwest::Request {
     fn host(&self) -> Option<String> {
```
Comment (thomassa) [inline src/reqwest_impls.rs:22]: Should we rename this fn to `host_and_port`?
Or does it need to match the behaviour of a `host` fn in `ClientRequest`? (I've not looked...)

Comment (robalar) [reply src/reqwest_impls.rs:22]: im not sure if that would be a breaking change or not? also it is for `Host` header so i guess the name is correct. I will add some docstrings to clarify its for the entire header not just the `host` fragment of the header

File: src/reqwest_impls.rs [hunk 4]
```diff
@@ -100,8 +109,90 @@ mod tests {
...
             "SHA-256=2vgEVkfe4d6VW+tSWAziO7BUx7uT/rA9hn1EoxUJi2o="
         );
+        assert_eq!(with_sig.headers().get(HOST).unwrap(), "test.com");
+    }
+
+    #[test]
+    fn it_works_blocking() {
+        let config = SigningConfig::new_default("test_key", "abcdefgh".as_bytes());
+
+        let client = reqwest::blocking::Client::new();
```
Comment (thomassa) [inline src/reqwest_impls.rs:119]: It took me a few minutes playing spot-the-difference to see that this line is the only difference between this test and the `it_works` test: I resorted to pasting the two of them into kdiff3 to display the difference.

Is it worth deduplicating?
Something like
```rust
    #[test]
    fn it_works_blocking() {
        it_works(reqwest::blocking::Client::new()
    }

    #[test]
    fn it_works_async() {
        it_works(reqwest::Client::new()
    }
```
or using a parameterisation library such as `test-case`.

It's the same for the other pair of tests: the ones specifying a non-standard port.

**Review by thomassa — APPROVED** [review]
I leave it to your judgement whether it's worth the time to factor out the duplication in the test cases.

Otherwise, all good.

File: src/reqwest_impls.rs (hunk 4 above)
Comment (robalar) [reply src/reqwest_impls.rs:119]: they are completely seperate types, and don't implement a shared trait for the builder (as far as i can tell) so I dont think its possible to construct a function that takes both as an argument?

**Pull request merged**